### DISCUSSION TITLE
How to work with @script & modules ?

### DISCUSSION THREAD

**Discussion by matfire** [livewire/livewire#8937] [1. Help]
Hello all,

I'm working on a project where I need to reference some node_modules "modules" from an Alpine data directive, but if I try to import the module from inside the <script> tag (wrapped in the @script tags), I get the following error: Uncaught SyntaxError: Cannot use import statement outside a module. Is Livewire doing something behind the scenes that changes how these things are actually shown in the dom ? Am I doing something wrong ?

I also tried creating a separate .js file, but the livewire events that I would need ("load") never get triggered :(

This is the offending section:
```blade
@script
<script type="module">
console.log("hello from module")
	import { basicSetup } from 'codemirror';
</script>
@endscript
```

**Comment by matfire** [comment]
Managed to make some progress;
using vite to load the script like this: `@vite('resources/js/editor.js')`
and using the following code:
```js

document.addEventListener('livewire:initialized', () => {
    const $wire = Livewire.first()
    Alpine.data('editor', () => {
        console.log('in editor')
        return {}
    })
})
```
correctly sets the $wire variable, but the Alpine.data either does not get called or is not set early enough :(

**Reply by matfire** [reply]
completely stepping of Alpine and going with some native js does not solve the issue either; when submitting some data using $wire.save() (save is the name of the method on the livewire side), all I had done with my js (i.e. load a CodeMirror editor) disappears...

**Comment by ImNIXX** [comment]
Hii @matfire 
We can solve this issue with dynamic import within a `<script type="module"></script>`. Dynamic import allows the module to be fetched asynchronously. This approach avoid any import related issue;

This is the solution -

```javascript
@script
  <script type="module">
    console.log("hello from module")
    (async () => {
      const { basicSetup  } = await import('codemirror');
      console.log(basicSetup  );
    })();
  </script>
@endscript
```

**Reply by matfire** [reply]
hey, thanks for the reply! I tried this, and I get the following error:
```
module.cjs.js:1848 Alpine Expression Error: console.log(...) is not a function

Expression: "console.log("hello from module")
    (async () => {
        const { basicSetup  } = await import('codemirror');
        console.log(basicSetup  );
    })();"

Failed to resolve module specifier 'codemirror'
```
I'm injecting Alpine myself but I don't think it should change something

**Comment by matfire** [comment]
Edit: I've also tried using `@assets` tag and importing from jsdelivr, but it does not work either (adding type="module" to them makes the different dependencies crash because they cannot find each other).

**Comment by matfire** [comment]
Ok, small update: managed to load some modules using esm.run, but then I get a lot of conflicts with multiple imports of the same module, but I'm certain this can work withing an Alpine data binding, at least...
```js
    Alpine.data('editor', () => {
        return {
            async init() {
                //const { EditorState, Compartment } = await import('https://esm.run/@codemirror/state');
                const { basicSetup, EditorView, ...rest  } = await import('https://esm.run/codemirror');
                //const { markdown } = await import('https://esm.run/@codemirror/lang-markdown');
                const { keymap } = await import('https://esm.run/@codemirror/view');
                const { defaultKeymap } = await import('https://esm.run/@codemirror/commands')
                //const { languages } = await import('https://esm.run/@codemirror/language-data');
                console.log('init alpine component')
                const baseCodeMirrorConfig = [basicSetup, EditorView.lineWrapping]
               /* let startState = EditorState.create({
                    doc: $wire.content,
                    extensions: [baseCodeMirrorConfig]
                })*/
                this.view = new EditorView(/*{state: startState}*/)
                $refs.editor.appendChild(this.view.dom)
            },
            "loading": false
        }
    })
```

the commented lines either don't work (because of some missing dependencies), or another package detects more than one instance of it (but this could probably just be a codemirror thing xD)

Some progress, but still a long way from working properly

**Reply by matfire** [reply]
I managed to run my script in a separate file (bundled with vite), and then communicate by sending events and making Alpine read values from data attributes. Problem is, it looks like the <script> tag gets refreshed every time I execute a function from Livewire (making the editor that gets loaded from that side reload every time).
